packages/common/src/hexToResourceId.ts

```diff
@@ -0,0 +1,30 @@
+import { Hex, hexToString, sliceHex } from "viem";
+import { ResourceId } from "./common";
+import { ResourceType, resourceTypes } from "./resourceTypes";
+import { resourceTypeIds } from "./resourceIdToHex";
+import { ReverseMap } from "./type-utils/common";
+
+const resourceTypeIdToType = Object.fromEntries(
+  Object.entries(resourceTypeIds).map(([key, value]) => [value, key])
+) as ReverseMap<typeof resourceTypeIds>;
+
+function getResourceType(resourceTypeId: string): ResourceType | undefined {
+  // TODO: replace Partial with `noUncheckedIndexedAccess`
+  const type = (resourceTypeIdToType as Partial<Record<string, ResourceType>>)[resourceTypeId];
+  if (resourceTypes.includes(type as ResourceType)) {
+    return type;
+  }
+}
+
+export function hexToResourceId(hex: Hex): ResourceId {
+  const namespace = hexToString(sliceHex(hex, 0, 14)).replace(/\0+$/, "");
+  const name = hexToString(sliceHex(hex, 14, 30)).replace(/\0+$/, "");
+  const resourceTypeId = hexToString(sliceHex(hex, 30, 32)).replace(/\0+$/, "");
+  const type = getResourceType(resourceTypeId);
+
+  if (!type) {
+    throw new Error(`Unknown resource type: ${resourceTypeId}`);
+  }
+
+  return { namespace, name, type };
```

Comment:
i would have expected `type` to be a reserved keyword in typescript, interesting that it's not